Source/JavaScriptCore/runtime/WaiterListManager.h

```diff
@@ -0,0 +1,235 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#pragma once
+
+#include "DeferredWorkTimer.h"
+#include "JSPromise.h"
+#include <wtf/Condition.h>
+#include <wtf/HashMap.h>
+#include <wtf/Lock.h>
+#include <wtf/SentinelLinkedList.h>
+
+namespace JSC {
+
+enum class AtomicsWaitType : uint8_t { Sync, Async };
+enum class AtomicsWaitValidation : uint8_t { Pass, Fail };
+
+class Waiter final : public WTF::BasicRawSentinelNode<Waiter>, public ThreadSafeRefCounted<Waiter> {
+    WTF_MAKE_FAST_ALLOCATED;
+
+public:
+    Waiter(VM* vm)
+        : m_vm(vm)
+        , m_isAsync(false)
+    {
+    }
+
+    Waiter(JSPromise* promise)
+        : m_vm(&promise->vm())
+        , m_ticket(m_vm->deferredWorkTimer->addPendingWork(*m_vm, promise, { }))
+        , m_isAsync(true)
+    {
+    }
+
+    bool isAsync() const
+    {
+        return m_isAsync;
+    }
+
+    VM* vm()
+    {
+        return m_vm;
+    }
+
+    void setVM(VM* vm)
+    {
+        m_vm = vm;
+    }
+
+    void nullOutVM()
+    {
+        m_vm = nullptr;
+    }
+
+    Condition& condition()
+    {
+        ASSERT(!m_isAsync);
+        return m_condition;
+    }
+
+    DeferredWorkTimer::Ticket ticket() const
+    {
+        ASSERT(m_isAsync);
+        return m_ticket;
+    }
+
+    DeferredWorkTimer::Ticket nullOutTicket()
+    {
+        ASSERT(m_isAsync);
+        return std::exchange(m_ticket, nullptr);
+    }
+
+    void setTimer(Ref<RunLoop::DispatchTimer>&& timer)
+    {
+        ASSERT(m_isAsync);
+        m_timer = WTFMove(timer);
+    }
+
+    bool hasTimer()
+    {
+        return !!m_timer;
+    }
+
+    void cancelTimer()
+    {
+        ASSERT(m_isAsync);
+        // If the timeout for AsyncWaiter is infinity, we won't dispatch any timer.
+        if (!m_timer)
+            return;
+        m_timer->stop();
+        m_timer = nullptr;
+    }
+
+private:
+    VM* m_vm { nullptr };
+    DeferredWorkTimer::Ticket m_ticket { nullptr };
+    RefPtr<RunLoop::DispatchTimer> m_timer { nullptr };
+    Condition m_condition;
+    bool m_isAsync { false };
+};
+
+class WaiterList : public ThreadSafeRefCounted<WaiterList> {
+    WTF_MAKE_FAST_ALLOCATED;
+
+public:
+    ~WaiterList()
+    {
+        removeIf([](Waiter*) {
+            return true;
+        });
+    }
+
+    void addLast(const AbstractLocker&, Waiter& waiter)
+    {
+        m_waiters.append(&waiter);
+        waiter.ref();
+        m_size++;
+    }
+
+    Ref<Waiter> takeFirst(const AbstractLocker&)
+    {
+        return take(*m_waiters.begin());
+    }
+
+    bool findAndRemove(const AbstractLocker&, Waiter& target)
+    {
+        for (auto iter = m_waiters.begin(); iter != m_waiters.end(); ++iter) {
+            if (&*iter == &target) {
+                removeWithUpdate(target);
+                return true;
+            }
+        }
+        return false;
+    }
```

Comment:
This looks wrong design. The best part of linked-list is `O(1)` removal. So we should not do `O(N)` removal here.
```
bool findAndRemove(const AbstractLocker&, Waiter& target)
{
    if (!target.isOnList())
        return false;
    removeWithUpdate(target);
    return true;
}
```

Reply:
You can add test to ensure that this entry is on this list, but we should not do it in release build.
```
#if ASSERT_ENABLED
if (target.isOnList()) {
    bool found = false;
    for (auto iter = m_waiters.begin(); iter != m_waiters.end(); ++iter) {
        if (&*iter == &target)
            found = true;
    }
    ASSERT(found);
}
#endif
```